crates/zkevm_circuits/src/bn254/ec_add/mod.rs

```diff
@@ -0,0 +1,363 @@
+
+use super::*;
+
+use self::ec_mul::implementation::{
+    convert_field_element_to_uint256, convert_uint256_to_field_element,
+};
+use self::input::EcAddCircuitInstanceWitness;
+
+pub mod implementation;
+pub mod input;
+
+pub const MEMORY_QUERIES_PER_CALL: usize = 4;
+pub const NUM_MEMORY_READS_PER_CYCLE: usize = 4;
+const EXCEPTION_FLAGS_ARR_LEN: usize = 6;
+
+#[derive(Derivative, CSSelectable)]
+#[derivative(Clone, Debug)]
+pub struct EcAddPrecompileCallParams<F: SmallField> {
+    pub input_page: UInt32<F>,
+    pub input_offset: UInt32<F>,
+    pub output_page: UInt32<F>,
+    pub output_offset: UInt32<F>,
+}
+
+impl<F: SmallField> EcAddPrecompileCallParams<F> {
+    pub fn from_encoding<CS: ConstraintSystem<F>>(_cs: &mut CS, encoding: UInt256<F>) -> Self {
+        let input_offset = encoding.inner[0];
+        let output_offset = encoding.inner[2];
+        let input_page = encoding.inner[4];
+        let output_page = encoding.inner[5];
+
+        let new = Self {
+            input_page,
+            input_offset,
+            output_page,
+            output_offset,
+        };
+
+        new
+    }
+}
+
+fn ecadd_precompile_inner<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    x1: &UInt256<F>,
+    y1: &UInt256<F>,
+    x2: &UInt256<F>,
+    y2: &UInt256<F>,
+) -> (Boolean<F>, (UInt256<F>, UInt256<F>)) {
+    let base_field_params = &Arc::new(bn254_base_field_params());
+
+    // We need to check for infinity prior to potential masking coordinates.
+    let point1_is_infinity = is_affine_infinity(cs, (x1, y1));
+    let point2_is_infinity = is_affine_infinity(cs, (x2, y2));
+
+    // Coordinates are masked with zero in-place if they are not in field.
+    let mut coordinates = ArrayVec::from([*x1, *y1, *x2, *y2]);
+    let coordinates_are_in_field = validate_in_field(cs, &mut coordinates, base_field_params);
+    let [x1, y1, x2, y2] = coordinates.into_inner().unwrap();
+
+    let x1 = convert_uint256_to_field_element(cs, &x1, base_field_params);
+    let y1 = convert_uint256_to_field_element(cs, &y1, base_field_params);
+
+    let point1_on_curve = is_on_curve(cs, (&x1, &y1), base_field_params);
+    let point1_is_valid = point1_on_curve.or(cs, point1_is_infinity);
+
+    // Mask the point with zero in case it is not on curve.
+    let zero = BN256SWProjectivePoint::zero(cs, base_field_params);
+    let unchecked_point = BN256SWProjectivePoint::from_xy_unchecked(cs, x1, y1);
+    let mut point1 =
+        BN256SWProjectivePoint::conditionally_select(cs, point1_on_curve, &unchecked_point, &zero);
+
+    let x2 = convert_uint256_to_field_element(cs, &x2, base_field_params);
+    let y2 = convert_uint256_to_field_element(cs, &y2, base_field_params);
+
+    let point2_on_curve = is_on_curve(cs, (&x2, &y2), base_field_params);
+    let point2_is_valid = point2_on_curve.or(cs, point2_is_infinity);
+
+    let mut result = projective_add(cs, &mut point1, (x2, y2));
```

Comment:
inside fn projective_add second point still, go through checking 
    // let x_is_zero = x.is_zero(cs);
    // let y_is_zero = y.is_zero(cs);

    // let is_point2_inf = x_is_zero.and(cs, y_is_zero);
    We already did it earlier ( optimization idea)